Heatmap Layer

balhar-jakub opened the pull request:
### Description of the Change

Provide the implementation of the HeatMap as a Tiled Layer. 

### Why Should This Be In Core?

It is highly desirable to provide the developers simple solution for showing heatmaps. 

### Benefits

Simple way to display data using the heatmap implementation. 

### Potential Drawbacks

More code to be managed in the core. 

### Applicable Issues

PJHogan commented:
Plenty of real-world use for this. Is there and ESA project helping to drive it? We want ESA virtual globe (VG) efforts being satisfied, knowing which projects are happily orbiting with ESA-NASA WorldWind and what more they might need driving our dev activity would be good to know. Working more hand-in-glove, ESA projects with WorldWind, should help greater uptake of this VG platform. Ideally, getting the world onto one-page for spatial data would greatly facilitate dynamics for global collaboration. WorldWind well wedded to Copernicus Sentinel data will help turn that data into pure gold! So Rapunzel WorldWind let down your hair so Prince Copernicus might climb away!

balhar-jakub commented:
I have updated the current version to contain the performance effective approach for ESA use cases. I also fixed some more minor issues around antemeridian.

balhar-jakub commented:
I have been working in the last two weeks on the version using Inverse Distance Weighting and there are two major issues I don't yet have a solution for:
1) Speed - I ddin't yet find a way to make IDW as fast as the current model. The slowing down is very visible especially in the case of the data relevant for Sentinel App. The code which tests this is in the Sentinel App repository. 
2) Tiling issues - I haven't yet found issue which would work well with the current tiling schema, but here I believe that I will find one. though the question is what will be the performance issues.

AkeluX commented on src/layer/heatmap/HeatMapLayer.js:
```diff
@@ -0,0 +1,410 @@
+define([
```
This file is missing the copyright header.

balhar-jakub replied:
Done

AkeluX commented on package-lock.json:
```diff
@@ -156,7 +156,7 @@
     "babylon": {
```
Are any of the changes in package-lock.json required?

balhar-jakub replied:
Removed

AkeluX replied:
I still see changes here. If they are not needed, please revert them.

AkeluX commented on examples/HeatMap.js:
```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2015-2018 WorldWind Contributors
```
Could you apply the same header as in https://github.com/NASAWorldWind/WebWorldWind/blob/develop/src/WorldWind.js to all files?

balhar-jakub replied:
Done

AkeluX replied:
This particular file still has the old copyright header. Please update it.

AkeluX commented on src/WorldWind.js:
```diff
@@ -765,6 +777,7 @@ define([ // PLEASE KEEP ALL THIS IN ALPHABETICAL ORDER BY MODULE NAME (not direc
         WorldWind['ClickRecognizer'] = ClickRecognizer;
         WorldWind['ColladaLoader'] = ColladaLoader;
         WorldWind['Color'] = Color;
+        WorldWind['ColoredTile'] = ColoredTile;
```
All new heatmap classes should be prefixed with HeatMap.

AkeluX commented on src/geom/MeasuredLocation.js:
```diff
@@ -0,0 +1,89 @@
+     * @param measure {Number} The measure associated with the location
+     * @constructor
+     */
+    var MeasuredLocation = function (latitude, longitude, measure) {
```
This class is already in develop. Please update.

AkeluX commented on src/layer/heatmap/IntervalType.js:
```diff
@@ -0,0 +1,15 @@
+     */
+    var IntervalType = {
+        CONTINUOUS: 0,
+        QUANTILES: 1
```
Could you describe how these two options differ and why they should both be in this first version?

AkeluX replied:
@balhar-jakub, as discussed, please add comments about what these two options do.

AkeluX commented on examples/HeatMap.js:
```diff
@@ -0,0 +1,60 @@
+            );
+        }
+
+        //
```
Remove empty comment line or complete it.

AkeluX commented on src/layer/heatmap/HeatMapColoredTile.js:
```diff
@@ -0,0 +1,91 @@
+    './HeatMapTile'
+], function (HeatMapTile) {
+    /**
+     * It provides us with a HeatMapTile available to HeatMapColoredTile.
```
Please remove all "It" in accordance with the contributor guidelines.

> Use present tense in all documentation. Examples are: "Indicates that ...", "Computes ...", "Returns ...".

AkeluX commented:
This is also missing the run configuration for the example, see #774 for the new structure.

AkeluX commented on examples/HeatMap.html:
```diff
@@ -0,0 +1,46 @@
+<!DOCTYPE html>
```
There are subtle differences between this file and the latest structure of the examples, e.g. Web World Wind versus Web WorldWind and the declaration of the canvas. Please review it thoroughly. You can use https://github.com/NASAWorldWind/WebWorldWind/blob/develop/examples/GeoJSON.html as reference.

AkeluX commented on examples/HeatMap.js:
```diff
@@ -0,0 +1,60 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+requirejs(['./WorldWindShim',
```
Please also review this one based on https://github.com/NASAWorldWind/WebWorldWind/blob/develop/examples/DeepPicking.js

AkeluX commented:
@balhar-jakub thanks for the latest changes. I've just fixed this displacement of the shape when zooming in or our.
